source/multigrid/mg_transfer_internal.cc

```diff
@@ -191,25 +197,60 @@ namespace internal
             tria->level_ghost_owners();
           std::map<int, std::vector<DoFPair>> send_data;
 
-          // * find owners of the level dofs and insert into send_data
-          // accordingly
-          for (const auto &dofpair : send_data_temp)
+          sort(send_data_temp.begin(),
+               send_data_temp.end(),
+               [](const auto &lhs, const auto &rhs) {
+                 if (lhs.level < rhs.level)
+                   return true;
+                 if (lhs.level > rhs.level)
+                   return false;
+
+                 if (lhs.level_dof_index < rhs.level_dof_index)
+                   return true;
+                 if (lhs.level_dof_index > rhs.level_dof_index)
+                   return false;
+
+                 if (lhs.global_dof_index < rhs.global_dof_index)
+                   return true;
+                 else
+                   return false;
+               });
+          send_data_temp.erase(unique(send_data_temp.begin(),
+                                      send_data_temp.end()),
+                               send_data_temp.end());
+
+          for (unsigned int level = 0; level < n_levels; ++level)
             {
-              std::set<types::subdomain_id>::iterator it;
-              for (it = neighbors.begin(); it != neighbors.end(); ++it)
-                {
-                  if (mg_dof
-                        .locally_owned_mg_dofs_per_processor(dofpair.level)[*it]
-                        .is_element(dofpair.level_dof_index))
-                    {
-                      send_data[*it].push_back(dofpair);
-                      break;
-                    }
-                }
-              // Is this level dof not owned by any of our neighbors? That would
-              // certainly be a bug!
-              Assert(it != neighbors.end(),
-                     ExcMessage("could not find DoF owner."));
+              const IndexSet &is_local = mg_dof.locally_owned_mg_dofs(level);
+
+              std::vector<unsigned int> level_dof_indices;
+              std::vector<unsigned int> global_dof_indices;
+              for (const auto &dofpair : send_data_temp)
+                if (dofpair.level == level)
+                  {
+                    level_dof_indices.push_back(dofpair.level_dof_index);
+                    global_dof_indices.push_back(dofpair.global_dof_index);
+                  }
+
+              IndexSet is_ghost(is_local.size());
+              is_ghost.add_indices(level_dof_indices.begin(),
+                                   level_dof_indices.end());
+
+              AssertThrow(level_dof_indices.size() == is_ghost.n_elements(),
+                          ExcMessage("Size does not match!"));
+
+              auto array =
```

Comment:
same as above

Reply:
@bangerth: thank you for all your comments and your feedback!

As a background info: I was struggling quite a while to get the code working if multiple tests are running in parallel (as is done in CI). MPI-tests (if I do not run multiple tests in parallel) have been working for quite a while (and the code was used in a full-machine run on SuperMUC-NG). I have figured out that the usage of `IBarrier` has somehow problems in context of the tests. If I replace this with my own implementation, which mimics the functionality of `IBarrier` all tests are passed, as CI is showing. The issue with the `IBarrier` made me to use the `#ifdef COMPUTE_INDEX_OWNER_USE_OWN_IBARRIER` to be able to identify the problem. This and the random tags are features, which will be removed.

Currently I am (locally) refactoring the code and tackling the `IBarrier` issue by providing an alternative implementation. The alternative implementation is also needed since `IBarrier` is a MPI 3.0 feature.

In the re-factored code, responsibilities will be clearly separated (and the code will be documented). There will be a class implementing a consensus algorithm (responsible for the communication pattern - see also https://htor.inf.ethz.ch/publications/img/hoefler-dsde-protocols.pdf) and class providing the information needed to compute the index owners. 

The advantage of splitting up the class is that we can reuse the consensus class in other instances, where the communication complexity does not scale optimally, e.g. `Utilities::MPI::compute_point_to_point_communication_pattern()`. But this is an issue for a future merge request.

Reply:
Thanks for the explanation. I assumed something like this, though it is of course annoying that a standard MPI feature doesn't work reliably.

I appreciate you're willing to re-factor the code and take my comments into account! It is certainly a cool feature that we'd like to use in other contexts as well!